snowplow-common-internal/kinesis/src/main/scala/com.snowplowanalytics.snowplow/sources/KinesisSource.scala

```diff
@@ -0,0 +1,252 @@
+                // We need to release the semaphore, so that fs2-aws handles checkpointing the end of the shard.
+                Logger[F].warn(
+                  s"Checkpointing failed on last record in shard. Ignoring error and instead try checkpointing of the shard end"
+                ) *>
+                  Sync[F].delay(metadata.lastRecordSemaphore.release())
+
+              case _: IllegalArgumentException if metadata.lastRecordSemaphore.availablePermits === 0 =>
+                // See https://github.com/snowplow/enrich/issues/657 and https://github.com/snowplow/enrich/pull/658
+                // This can happen near the shard end, e.g. the penultimate batch in the shard, when KCL has already enqueued the final record in the shard to the fs2 queue.
+                // We must not release the semaphore yet, because we are not ready for fs2-aws to checkpoint the end of the shard.
+                // We can safely ignore the exception and move on.
+                Logger[F].warn(
+                  s"Checkpointing failed on a record which was not the last in the shard. Meanwhile, KCL has already enqueued the final record in the shard to the fs2 queue. Ignoring error and instead continue processing towards the shard end"
+                )
+            }
+        }
+    def nack(c: List[KinesisMetadata[F]]): F[Unit] = Applicative[F].unit
+  }
+
+  private def kinesisStream[F[_]: Async](config: KinesisSourceConfig): Stream[F, LowLevelEvents[List[KinesisMetadata[F]]]] = {
+    val region = Region.of(config.region.getOrElse(getRuntimeRegion))
+    val resources =
+      for {
+        consumerSettings <- Resource.pure[F, KinesisConsumerSettings](
+                              KinesisConsumerSettings(
+                                config.streamName,
+                                config.appName,
+                                region,
+                                bufferSize = PosInt.unsafeFrom(config.bufferSize)
+                              )
+                            )
+        kinesisClient <- mkKinesisClient[F](region, config.customEndpoint)
+        dynamoClient <- mkDynamoDbClient[F](region, config.dynamodbCustomEndpoint)
+        cloudWatchClient <- mkCloudWatchClient[F](region, config.cloudwatchCustomEndpoint)
+        kinesis <- Resource.pure[F, Kinesis[F]](
+                     Kinesis.create(scheduler(kinesisClient, dynamoClient, cloudWatchClient, config, _))
+                   )
+      } yield (consumerSettings, kinesis)
+
+    Stream
+      .resource(resources)
+      .flatMap { case (settings, kinesis) =>
+        kinesis.readFromKinesisStream(settings)
+      }
+      .chunks
+      .map { chunk =>
+        LowLevelEvents(chunk.toList.map(getPayload), chunk.toList.map(toMetadata))
+      }
+  }
+
+  def getPayload(record: CommittableRecord): Array[Byte] = {
+    val data = record.record.data
+    val buffer = ArrayBuffer[Byte]()
+    while (data.hasRemaining)
+      buffer.append(data.get)
+    buffer.toArray
```

Comment:
Just pointing out that [rdb transformer has a slightly different implementation](https://github.com/snowplow/snowplow-rdb-loader/blob/5.7.3/modules/aws/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/aws/Kinesis.scala#L96-L97)

Which is better?  I _guess_ it's the rdb-transformer implementation, because it copies in bulk instead of byte-by-byte.  The difference is important, because we'll do this for every single snowplow event.

**Idea** -  Both implementations are copying data from a ByteBuffer to an array.  This is expensive -- especially if we do it again and again for each event.  How about we set things up so that we don't need to copy the data!

We could change the definition of [TokenedEvents](https://github.com/snowplow-incubator/snowplow-lake-loader/blob/0.1.0/snowplow-common-internal/streams-core/src/main/scala/com.snowplowanalytics.snowplow/sources/TokenedEvents.scala#L22) so that `events` is a `List[ByteBuffer]`?  I think the downstream application can still work with that type, and it means we never need to copy bytes.

Reply:
I liked the idea and making the required changes, thanks!